discord/commands/mixins.py

```diff
@@ -0,0 +1,910 @@
+from __future__ import annotations
+
+import asyncio
+import datetime
+import functools
+from typing import TYPE_CHECKING, Any, Callable, Coroutine, Generic, TypeVar, Union
+
+from .. import abc, utils
+from ..errors import (
+    ApplicationCommandError,
+    CheckFailure,
+    CommandError,
+    CommandInvokeError,
+    CommandOnCooldown,
+    DisabledCommand,
+)
+from .cooldowns import BucketType, CooldownMapping, MaxConcurrency
+
+if TYPE_CHECKING:
+    from typing_extensions import ParamSpec
+
+    from ..abc import MessageableChannel
+    from ..bot import AutoShardedBot, Bot
+    from ..cog import Cog
+    from ..guild import Guild
+    from ..interactions import Interaction
+    from ..member import Member
+    from ..message import Message
+    from ..state import ConnectionState
+    from ..user import ClientUser, User
+    from ..voice_client import VoiceProtocol
+
+    P = ParamSpec("P")
+else:
+    P = TypeVar("P")
+
+BotT = TypeVar("BotT", bound="Union[Bot, AutoShardedBot]")
+CogT = TypeVar("CogT", bound="Cog")
+CallbackT = TypeVar("CallbackT")
+ContextT = TypeVar("ContextT", bound="BaseContext")
+
+T = TypeVar("T")
+Coro = Coroutine[Any, Any, T]
+MaybeCoro = Union[T, Coro[T]]
+
+Check = Union[
+    Callable[[CogT, ContextT], MaybeCoro[bool]],
+    Callable[[ContextT], MaybeCoro[bool]],
+]
+
+Error = Union[
+    Callable[[CogT, "BaseContext[Any]", CommandError], Coro[Any]],
+    Callable[["BaseContext[Any]", CommandError], Coro[Any]],
+]
+ErrorT = TypeVar("ErrorT", bound="Error")
+
+Hook = Union[Callable[[CogT, ContextT], Coro[Any]], Callable[[ContextT], Coro[Any]]]
+HookT = TypeVar("HookT", bound="Hook")
+
+
+__all__ = (
+    "Invokable",
+    "_BaseCommand",
+    "BaseContext",
+)
+
+
+def unwrap_function(function: Callable[..., Any]) -> Callable[..., Any]:
```

Comment:
Shouldn't you set the type hint of `function` to a union of a `functools.partial` and a generic `Callable`?